src/main/kotlin/org/wfanet/measurement/duchy/daemon/mill/liquidlegionsv2/LiquidLegionsV2Mill.kt

```diff
@@ -533,45 +535,49 @@ class LiquidLegionsV2Mill(
     return completeComputation(nextToken, CompletedReason.SUCCEEDED)
   }
 
+  // Gets the order of all duchies in the MPC ring. Since it is a ring, we put the aggregator at the
+  // end for simplicity.
+  private fun getMpcDuchyRing(token: ComputationToken): List<String> {
+    val nonAggregatorsSet = cryptoKeySet.allDuchyPublicKeys
+      .filter { it.key != aggregatorId }
+      .map { Duchy(it.key, it.value.element.toStringUtf8()) }.toSet()
+    return getDuchyOrderByPublicKeysAndComputationId(
+      nonAggregatorsSet, token.globalComputationId
+    ) + aggregatorId
+  }
+
   private fun nextDuchyStub(token: ComputationToken): ComputationControlCoroutineStub {
-    val nextDuchy = token.computationDetails.liquidLegionsV2.outgoingNodeId
+    val mpcDuchyRing = getMpcDuchyRing(token)
+    val nextDuchy = getNextDuchy(mpcDuchyRing, duchyId)
     return workerStubs[nextDuchy]
       ?: throw PermanentComputationError(
         IllegalArgumentException("No ComputationControlService stub for next duchy '$nextDuchy'")
       )
   }
 
-  private fun aggregatorDuchyStub(token: ComputationToken): ComputationControlCoroutineStub {
-    val aggregatorDuchy = token.computationDetails.liquidLegionsV2.aggregatorNodeId
-    return workerStubs[aggregatorDuchy]
+  private fun aggregatorDuchyStub(): ComputationControlCoroutineStub {
+    return workerStubs[aggregatorId]
       ?: throw PermanentComputationError(
         IllegalArgumentException(
-          "No ComputationControlService stub for primary duchy '$aggregatorDuchy'"
+          "No ComputationControlService stub for primary duchy '$aggregatorId'"
         )
       )
   }
 
   // Combines the public keys from the duchies after this duchy in the ring and the primary duchy.
-  fun getPartiallyCombinedPublicKey(globalId: String, nextDuchy: String): ElGamalPublicKey {
-    val order = duchyOrder.computationOrder(globalId)
-    require(order.contains(nextDuchy)) { "$nextDuchy doesn't exist in the duchy ring." }
-    val aggregatorId = order[0]
-    if (nextDuchy == aggregatorId) {
-      // Return the aggregator's key if next duchy is the aggregator.
-      return cryptoKeySet.otherDuchyPublicKeys[nextDuchy] ?: error(
-        "$nextDuchy is not in the key set."
-      )
+  fun getPartiallyCombinedPublicKey(token: ComputationToken): ElGamalPublicKey {
+    require(token.computationDetails.liquidLegionsV2.role == NON_AGGREGATOR) {
+      "only NON_AGGREGATOR should call getPartiallyCombinedPublicKey()."
     }
-
-    val partialList = order.subList(order.indexOf(nextDuchy), order.size) + aggregatorId
+    val partialList = getFollowingDuchies(getMpcDuchyRing(token), duchyId)
     val lookupKey = partialList.sorted().joinToString()
 
     return partiallyCombinedPublicKeyMap.computeIfAbsent(lookupKey) {
```

Comment:
What about storing the combined public key for following duchies in the ComputationDetails? It would save a few RPCs and calculations per computation, but would not be that much in the grand scheme of things I suppose. Thoughts?

Reply:
they are already cached in memory. see line 575 "partiallyCombinedPublicKeyMap.computeIfAbsent()"

However, you actually pointed out a bug (for the future) here. We are going to use different keys for different computations, so each result is only used once, no need to cache at all. If we forget to delete the partiallyCombinedPublicKeyMap, a wrong key would be used for following computations.

Reply:
oops, you are saying storing them in the computationDetails.

We don't need to store it anywhere when new keys are used in different computations. Each partially combined key will be only used once at that computation (at execution phase two).